homeassistant/components/fritzbox_callmonitor/config_flow.py

```diff
@@ -0,0 +1,233 @@
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_PORT, CONF_USERNAME
+from homeassistant.core import callback
+
+from .base import FritzBoxPhonebook
+
+# pylint:disable=unused-import
+from .const import (
+    CONF_PHONEBOOK,
+    CONF_PREFIXES,
+    DEFAULT_HOST,
+    DEFAULT_PHONEBOOK,
+    DEFAULT_PORT,
+    DEFAULT_USERNAME,
+    DOMAIN,
+    FRITZ_ACTION_GET_INFO,
+    FRITZ_ATTR_NAME,
+    FRITZ_ATTR_SERIAL_NUMBER,
+    FRITZ_SERVICE_DEVICE_INFO,
+    SERIAL_NUMBER,
+)
+
+DATA_SCHEMA_USER = vol.Schema(
+    {
+        vol.Required(CONF_HOST, default=DEFAULT_HOST): str,
+        vol.Required(CONF_PORT, default=DEFAULT_PORT): vol.Coerce(int),
+        vol.Required(CONF_USERNAME, default=DEFAULT_USERNAME): str,
+        vol.Required(CONF_PASSWORD): str,
+    }
+)
+
+RESULT_INVALID_AUTH = "invalid_auth"
+RESULT_INSUFFICIENT_PERMISSIONS = "insufficient_permissions"
+RESULT_NO_DEVIES_FOUND = "no_devices_found"
+RESULT_SUCCESS = "success"
+
+
+class FritzBoxCallMonitorConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a fritzbox_callmonitor config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize flow."""
+        self._name = None
+        self._host = None
+        self._port = None
+        self._username = None
+        self._password = None
+        self._phonebook_name = None
+        self._phonebook_id = None
+        self._phonebook_ids = None
+        self._phonebook = None
+        self._prefixes = None
+        self._serial_number = None
+
+    def _get_entry(self):
+        """Create and return an entry."""
+        return self.async_create_entry(
+            title=self._phonebook_name,
+            data={
+                CONF_HOST: self._host,
+                CONF_PORT: self._port,
+                CONF_USERNAME: self._username,
+                CONF_PASSWORD: self._password,
+                CONF_PHONEBOOK: self._phonebook_id,
+                CONF_PREFIXES: self._prefixes,
+                SERIAL_NUMBER: self._serial_number,
+            },
+        )
+
+    def _try_connect(self):
+        """Try to connect and check auth."""
+        self._phonebook = FritzBoxPhonebook(
+            host=self._host,
+            username=self._username,
+            password=self._password,
+            phonebook_id=self._phonebook_id,
+            prefixes=self._prefixes,
+        )
+
+        try:
+            self._phonebook.init_phonebook()
+            self._phonebook_ids = self._phonebook.get_phonebook_ids()
+
+            fritz_connection = FritzConnection(
+                address=self._host, user=self._username, password=self._password
+            )
+            device_info = fritz_connection.call_action(
+                FRITZ_SERVICE_DEVICE_INFO, FRITZ_ACTION_GET_INFO
+            )
+            self._serial_number = device_info[FRITZ_ATTR_SERIAL_NUMBER]
```

Comment:
Not sure if this is the correct spot to do this, however it can throw the same exceptions so I thought it would be best suited here. Unfortunately, we have to create a separate object and cannot use FritzPhoneBook to access the serial number of the FritzBox.

Reply:
This is fine.